clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java

```diff
@@ -0,0 +1,473 @@
+        } else {
+            clear(key);
+            future.complete(value);
+        }
+    }
+
+    /**
+     * Check whether any requests need to be sent. This should be called immediately
+     * after the driver is constructed and then again after each request returns
+     * (i.e. after {@link #onFailure(long, RequestSpec, Throwable)} or
+     * {@link #onResponse(long, RequestSpec, AbstractResponse)}).
+     *
+     * @return A list of requests that need to be sent
+     */
+    public List<RequestSpec<K>> poll() {
+        List<RequestSpec<K>> requests = new ArrayList<>();
+        collectLookupRequests(requests);
+        collectFulfillmentRequests(requests);
+        return requests;
+    }
+
+    /**
+     * Get a map of the futures that are awaiting completion.
+     */
+    public Map<K, KafkaFutureImpl<V>> futures() {
+        return futures;
+    }
+
+    /**
+     * Callback that is invoked when a `Call` returns a response successfully.
+     */
+    public void onResponse(
+        long currentTimeMs,
+        RequestSpec<K> spec,
+        AbstractResponse response
+    ) {
+        clearInflightRequest(currentTimeMs, spec);
+
+        if (spec.scope instanceof FulfillmentScope) {
+            int brokerId = ((FulfillmentScope) spec.scope).destinationBrokerId;
+            AdminApiHandler.ApiResult<K, V> result = handler.handleResponse(
+                brokerId,
+                spec.keys,
+                response
+            );
+            result.completedKeys.forEach(this::complete);
+            result.failedKeys.forEach(this::completeExceptionally);
+            result.unmappedKeys.forEach(this::unmap);
+        } else {
+            DynamicKeyMapping<K> mapping = dynamicMapping.orElseThrow(() ->
+                new IllegalStateException("Received response for unexpected dynamic lookup scope " + spec.scope)
+            );
+
+            AdminApiLookupStrategy.LookupResult<K> result = mapping.lookupStrategy.handleResponse(
+                spec.keys,
+                response
+            );
+            result.failedKeys.forEach(this::completeExceptionally);
+            result.mappedKeys.forEach(this::map);
+        }
+    }
+
+    /**
+     * Callback that is invoked when a `Call` is failed.
+     */
+    public void onFailure(
+        long currentTimeMs,
+        RequestSpec<K> spec,
+        Throwable t
+    ) {
+        clearInflightRequest(currentTimeMs, spec);
+        if (t instanceof DisconnectException) {
+            log.debug("Node disconnected before response could be received for request {}. " +
+                "Will attempt retry", spec.request);
+            dynamicMapping.ifPresent(mapping -> {
+                // After a disconnect, we want the driver to attempt to lookup the key
+                // again. This gives us a chance to find a new coordinator or partition
+                // leader for example.
+                if (mapping.keys.containsAll(spec.keys)) {
```

Comment:
This check is a bit mysterious to me as the dynamic keys can come from different sources. First source is from `dynamicMapping` and another is from `AdminApiHandler.ApiResult<K, V>` (when processing `FulfillmentScope`). What will happen if the keys consist of dynamic key and static key? For example, `AdminApiHandler#handleResponse` returns `unmappedKeys` with `static key`? Should it make quick failure if above case is not allowed?

Reply:
If I understand correctly, the concern is that we may have a case which includes both static and dynamic keys. I think that is fair. It would make more sense to check the keys individually. I do not think I have any test cases for this scenario, so let me add one.